### PR TITLE
PERF-1992 Upstream LinkStoreMongoDbBasic.java

### PR DESCRIPTION
**Pull request by henrikingo**
This is a basic port of LinkStoreMysql to MongoDB.
See LinkStoreMongoDb2 for a more optimized version.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -612,21 +617,36 @@ For other forms of authentication, please use the __url__ property. See
...
-    ./bin/linkbench -c config/LinkConfigMongoDb.properties -r
+    ./bin/linkbench -c config/LinkConfigMongoDb2.properties -r
 
 Running Tests
 -------------
 
 You can use the following to run some / all of the tests:
 
-    $> mvn test -Dtest=**/Mongo* test
+    $> mvn test -P mongodb-tests
```
Comment (mdcallag) [inline README.md:652]: Which tests run when you do that? AFAIK some of the MongoDB tests are tagged and won't be run by this.

File: src/main/java/com/facebook/LinkBench/LinkStoreMongoDbBasic.java [hunk 2]
```diff
@@ -0,0 +1,1280 @@
...
+            logger.warn("testCount: null total");
+        }
+
+        if (count != total) {
+            throw new CommandBlockException("Data inconsistency between " + assoctable + "(" + count + ") " +
+                    " and " + counttable + "(" + total + ")");
+        }
+    }
+
+    // not called anywhere, addLink is called directly in LinkBenchRequest
```
Comment (mdcallag) [inline src/main/java/com/facebook/LinkBench/LinkStoreMongoDbBasic.java:754]: The "not called anywhere" comment is incorrect now. Used to be true, but LinkBenchRequest.java calls it unless the "neverChange" option is set.

File: config/LinkConfigMongoDbBasic.properties [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+# etc to warm up.
+warmup_time = 0
+
+# seed for request random number generation (optional)
+# request_random_seed = 12345
+
+# maximum number of failures per requester to tolerate before aborting
+# negative number means never abort
+max_failed_requests = 100
+
```
Comment (mdcallag) [inline config/LinkConfigMongoDbBasic.properties:148]: If you add "never_change = true" then old behavior is preserved

File: README.md (hunk 1 above)
Comment (henrikingo) [reply README.md:652]: They are both identical:

    -------------------------------------------------------
     T E S T S
    -------------------------------------------------------
    parallel='none', perCoreThreadCount=true, threadCount=2, useUnlimitedThreads=false
    Running com.facebook.LinkBench.MongoDbGraphStoreTestBasic
    Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 99.002 sec
    Running com.facebook.LinkBench.MongoDbLinkStoreTestBasic
    Tests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 276.626 sec
    Running com.facebook.LinkBench.MongoDbLinkStoreTest2
    Tests run: 15, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 271.101 sec
    Running com.facebook.LinkBench.MongoDbNodeStoreTestBasic
    Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.4 sec
    Running com.facebook.LinkBench.MongoDbNodeStoreTest2
    Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.612 sec
    Running com.facebook.LinkBench.MongoDbGraphStoreTest2
    Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 127.484 sec

    Results :

    Tests run: 40, Failures: 0, Errors: 0, Skipped: 0

This is the profile, it's the same pattern: https://github.com/mdcallag/linkbench/blob/master/pom.xml#L132-L148

**Comment by henrikingo** [comment]
David is preparing for conference talks and passed this review.

**Pull request merged**